Rework introduction to SSA-IR in dev docs (#49796)
JuliaLang/julia: doc/src/devdocs/ssair.md

2. if/else and loops are turned into `goto` statements.
3. lines with multiple operations are split into multiple lines by introducing variables.

For example the following Julia code:
```julia
function foo(x)
    y = sin(x)
    if x > 5.0
        y = y + cos(x)
    end
    return exp(2) + y
end
```
when called with a `Float64` argument is translated into:

```julia
using InteractiveUtils
@code_typed foo(1.0)
```

```llvm
CodeInfo(
1 ─ %1 = invoke Main.sin(x::Float64)::Float64
│   %2 = Base.lt_float(x, 5.0)::Bool
└──      goto #3 if not %2
2 ─ %4 = invoke Main.cos(x::Float64)::Float64
└── %5 = Base.add_float(%1, %4)::Float64
3 ┄ %6 = φ (#2 => %5, #1 => %1)::Float64
│   %7 = Base.add_float(7.38905609893065, %6)::Float64
└──      return %7
) => Float64
```

In this example, we can see all of these changes.
1. The first basic block is everything in
```llvm
1 ─ %1 = invoke Main.sin(x::Float64)::Float64
│   %2 = Base.lt_float(x, 5.0)::Bool
└──      goto #3 if not %2
```
2. The `if` statement is translated into `goto #3 if not %2` which goes to the 3rd basic block if `x>5` isn't met and otherwise goes to the second basic block.

> 3. `%2` is a SSA value introduced to represent `x > 5`.

Reviewer: ```suggestion
3. `%2` is an SSA value introduced to represent `x > 5`.
```
minor nit: grammatically an `S` is pronounced `ess` phonetically, so it takes `an` as the article

# Julia SSA-form IR

Julia uses a static single assignment intermediate representation ([SSA IR](https://en.wikipedia.org/wiki/Static_single-assignment_form)) to perform optimization.
This IR is different from LLVM IR, and unique to Julia.
It allows for Julia specific optimizations.

> The SSA part means that in the IR, every variable is assigned to exactly once and this definition occurs before all uses of the variable. The transformation to IR (SSA conversion only happens after inference) is performed by code lowering, and in this representation there are a few changes worth highlighting.

Reviewer: ```suggestion
```
This sentence is mostly false. SSA is mostly a claim about reachability and side-effects of our instructions, and that every instruction has a unique numbering that identifies its side effects and result. Those facts are true and present prior to inference. No new SSA properties are added after inference.

Author: Thanks, it has grammatical errors as well.

2. if/else and loops are turned into `goto` statements.
3. lines with multiple operations are split into multiple lines by introducing variables.

For example the following Julia code:
```julia
function foo(x)
    y = sin(x)
    if x > 5.0
        y = y + cos(x)
    end
    return exp(2) + y
end
```
when called with a `Float64` argument is translated into:

```julia
using InteractiveUtils
@code_typed foo(1.0)
```

```llvm
CodeInfo(
1 ─ %1 = invoke Main.sin(x::Float64)::Float64
│   %2 = Base.lt_float(x, 5.0)::Bool
└──      goto #3 if not %2
2 ─ %4 = invoke Main.cos(x::Float64)::Float64
└── %5 = Base.add_float(%1, %4)::Float64
3 ┄ %6 = φ (#2 => %5, #1 => %1)::Float64
│   %7 = Base.add_float(7.38905609893065, %6)::Float64
└──      return %7
) => Float64
```

In this example, we can see all of these changes.
1. The first basic block is everything in
```llvm
1 ─ %1 = invoke Main.sin(x::Float64)::Float64
│   %2 = Base.lt_float(x, 5.0)::Bool
└──      goto #3 if not %2
```
2. The `if` statement is translated into `goto #3 if not %2` which goes to the 3rd basic block if `x>5` isn't met and otherwise goes to the second basic block.

> 3. `%2` is a variable introduced to represent `x > 5`.

Reviewer: `%2` is not a variable (it cannot change or vary). It is rather instead simultaneously a representation of the index of the expression (when printed on the LHS) and the value it produced (when used on the RHS).

Author: Would "SSA value" be correct?

Author: ```suggestion
3. `%2` is a SSA value introduced to represent `x > 5`.
```
